[stable/nginx-ingress] Add RBAC support

jsulinski opened the pull request:
Enable RBAC as per: https://github.com/kubernetes/ingress/tree/master/examples/rbac/nginx and https://github.com/kubernetes/ingress/issues/266

cknowles commented on stable/nginx-ingress/templates/controller-deployment.yaml:
```diff
@@ -22,6 +25,9 @@ spec:
     spec:
+{{- if .Values.controller.rbac.enabled }}
+      serviceAccountName: {{ template "fullname" . }}
+{{- end }}
```
I submitted a similar update for Traefik in https://github.com/kubernetes/charts/pull/1225. When switching RBAC enabled from true to false and helm installing, I found the serviceAccountName stays as the Traefik specific one rather than being reset to `default`. I wondered if you had encountered that issue here?

jsulinski replied:
No, I didn't experience this. I would presume the issue is elsewhere in your case, as this conditional logic is very simple.

cknowles replied:
@jsulinski hmmm strange the Traefik chart doesn't have any complex logic here, the switch is pretty much identical.

icereval replied:
Should this also have a `default` service account? https://github.com/kubernetes/charts/blob/master/stable/kube2iam/templates/daemonset.yaml#L24

icereval replied:
I believe you'll also want to apply this to the [`controller-daemonset.yaml`](https://github.com/kubernetes/charts/blob/master/stable/nginx-ingress/templates/controller-daemonset.yaml#L24).

jsulinski replied:
I agree. Pulled your changes.

linki commented on stable/nginx-ingress/templates/controller-deployment.yaml:
```diff
@@ -9,6 +9,9 @@ metadata:
     release: {{ .Release.Name }}
   name: {{ template "controller.fullname" . }}
+{{- if .Values.controller.rbac.enabled }}
+  namespace: {{ .Release.Namespace }}
```
This seems weird. I think we prefer to leave out the `namespace` in the metadata section of manifests and let helm take care of that when the `--namespace` flag is used.

However, you probably need it when you reference the `subject` in the role bindings, but it shouldn't be required here. Can you give it a try? (including the other manfiests)

jsulinski replied:
Updated, thanks so much! You were spot on.

bison commented on stable/nginx-ingress/templates/controller-role.yaml:
```diff
@@ -0,0 +1,37 @@
+      # Here: "<ingress-controller-leader>-<nginx>"
+      # This has to be adapted if you change either parameter
+      # when launching the nginx-ingress-controller.
+      - "ingress-controller-leader-nginx"
```
This seems like it should be a user-supplied list defaulting to `["ingress-controller-leader-nginx"]`.

jsulinski replied:
Agreed. Will update.

bison commented:
This looks great. Hope it's merged soon. I think the example RBAC rules [here](https://github.com/kubernetes/ingress/blob/master/examples/rbac/nginx/nginx-ingress-controller-rbac.yml) have been updated in the last few days. Might want to pull in the latest version.

flah00 commented:
<strike>@jsulinski I have been testing this PR with the latest nginx-ingress-controller 0.9.0 beta and found that I had to add create, delete, and patch for configmaps. It seems that the controller maintains a configmap named `ingress-controller-leader-nginx`.</strike>

Disregard, I made a mistake, when I merged your branch ... derp.

jamielennox commented on stable/nginx-ingress/values.yaml:
```diff
@@ -36,6 +36,10 @@ controller:
 
+  ## enable RBAC as per: https://github.com/kubernetes/ingress/tree/master/examples/rbac/nginx and https://github.com/kubernetes/ingress/issues/266
+  rbac:
+    enabled: false
```
nit: The two charts with optional RBAC support that have merged seem to be stable/etcd-operator and incubator/istio and both use .Values.rbac.install as the flag. Personally i prefer enabled, but thought i'd point it out for consistency.

jsulinski replied:
Great point. I'll update when I have a free moment.

icereval replied:
fyi, looks like the Traefik chart is merged w/ `rbac.enabled`, :+1:

https://github.com/kubernetes/charts/blob/master/stable/traefik/values.yaml#L63

jsulinski replied:
Sticking with rbac.enabled.

ReSearchITEng commented:
Once merged, it will solve: #927

icereval commented on stable/nginx-ingress/templates/controller-cluster-role.yaml:
```diff
@@ -0,0 +1,54 @@
+  - apiGroups:
+      - ""
+    resources:
+        - events
```
nit; spacing is off just a bit

jsulinski replied:
Thanks!

icereval commented on stable/nginx-ingress/templates/controller-cluster-role.yaml:
```diff
@@ -0,0 +1,54 @@
+  - apiGroups:
+      - ""
+    resources:
+      - nodes
```
when using `--watch-namespace` I had to add `namespaces` here to avoid the following error:

```bash
F0710 18:20:42.008164      10 launch.go:145] no watchNamespace with name stage-1-osf found: User "system:serviceaccount:stage-1-osf:mfjzdy-nginx-ingress" cannot get namespaces in the namespace "stage-1-osf".: "Unknown user \"system:serviceaccount:stage-1-osf:mfjzdy-nginx-ingress\"" (get namespaces stage-1-osf)
```

jsulinski replied:
Can you elaborate on this?

icereval replied:
The controller is attempting to request a list of namespaces and due to lack of permissions throws an exception `(get namespaces stage-1-osf)`. This command option essentially allows you to have the controller only watch a specified namespace for ingress' tagged with `nginx`, rather than a single global ingress nginx controller for the entire cluster.

In our case we are evaluating dedicated ip/ingress for each significant component of a larger application, thinking more towards a potential future of adding ModSecurity at this controller ingress point.

jsulinski replied:
Can you share an example? I haven't used --watch-namespace yet.

icereval replied:
```
                             .─────────────────.                              
                         _.─'                   `──.                          
                        (         Internet          )                         
                         `───.                 _.──'                          
                              `───────────────'                               
                                      │                                       
                                      ▼                                       
                          ┌───────────────────────┐                           
                          │  Google Load Balancer │                           
                          └───────────────────────┘                           
                                      │                                       
                    ┌─────────────────┴──────────────────┐                    
┌─Cluster #1────────┼────────────────────────────────────┼───────────────────┐
│                   ▼                                    ▼                   │
│  ┌Namespace (Testing)──────────────┐  ┌Namespace (Staging)──────────────┐  │
│  │┌──────────────┐ ┌──────────────┐│  │┌──────────────┐ ┌──────────────┐│  │
│  ││   Ingress    │ │   Default    ││  ││   Ingress    │ │   Default    ││  │
│  ││  Controller  │ │   Backend    ││  ││  Controller  │ │   Backend    ││  │
│  │└──────────────┘ └──────────────┘│  │└──────────────┘ └──────────────┘│  │
│  │        │                ▲       │  │        │                ▲       │  │
│  │        └───────┬────────┘       │  │        └───────┬────────┘       │  │
│  │                │                │  │                │                │  │
│  │       ┌────────────────┐        │  │       ┌────────────────┐        │  │
│  │       │  ingress.yaml  │        │  │       │  ingress.yaml  │        │  │
│  │       └────────────────┘        │  │       └────────────────┘        │  │
│  │                │                │  │                │                │  │
│  │     ┌──────────┼──────────┐     │  │     ┌──────────┼──────────┐     │  │
│  │     ▼          ▼          ▼     │  │     ▼          ▼          ▼     │  │
│  │ ┌───────┐  ┌───────┐  ┌───────┐ │  │ ┌───────┐  ┌───────┐  ┌───────┐ │  │
│  │ │ Pod 1 │  │ Pod 2 │  │ Pod 3 │ │  │ │ Pod 1 │  │ Pod 2 │  │ Pod 3 │ │  │
│  │ └───────┘  └───────┘  └───────┘ │  │ └───────┘  └───────┘  └───────┘ │  │
│  └─────────────────────────────────┘  └─────────────────────────────────┘  │
│                                                                            │
└────────────────────────────────────────────────────────────────────────────┘
```

viglesiasce replied:
please tell me how you made that diagram! (programatically i hope)

jsulinski replied:
Thank you! I'll work on this as soon as I have a moment.

icereval replied:
@viglesiasce, using this awesome tool https://monodraw.helftone.com/

jsulinski commented:
I'm working on testing revisions but having some issues with my RBAC enabled Kube cluster and chart. I'll push the changes without testing if I can't make progress soon.

icereval committed:
Add RBAC support for nginx-ingress

based on: https://github.com/kubernetes/charts/pull/1235

jsulinski committed:
nginx-ingress: fix spacing for events

jsulinski commented:
This is tested with all changes except making resourceNames a value. I'll do that as soon as I can, or anyone else is welcome to.

jsulinski commented:
Thanks @icereval!

jsulinski committed:
Pull in new RBAC changes from https://github.com/kubernetes/ingress/commit/4618fd2f64a904b1949f1d0a9a76ebe6ab8cb719

jsulinski commented:
@bison Thanks, updated.

jsulinski committed:
Move resourceNames to a value

Commit:
templates: use a range loop to render the resourceNames

* this produces the correctly formatted output we are looking for, specifically
  with regard to the extra carriage return which didn't want to be chomped away
  by the usual `-` method.

jsulinski commented:
That should be everything except the watchNamespace functionality.

mgoodness commented:
Something to consider is that a user may need to use a non-`default` ServiceAccount that isn't created/managed by Helm. Any thoughts on killing two birds with one stone (maybe like [this](https://github.com/coreos/alb-ingress-controller/pull/146))?

jsulinski commented:
I like this idea.

mgoodness commented:
@jsulinski Would you like me to wrap this up? You're 90% finished, and I can apply the changes I mentioned very quickly & easily. Just don't want to step on your toes.

The only hitch is that I'd like to move `controller.rbac.enabled` to `rbac.create`, as it is [here](https://github.com/coreos/alb-ingress-controller/pull/146) and [here](https://github.com/kubernetes/charts/pull/1504). I know you'd settled on `enabled`, but I think `create` more accurately reflects what it does.

jsulinski commented:
That would be awesome! Don't worry about stepping on my toes, just trying to contribute back when I have the opportunity to do so. 

I don't feel strongly at all about enabled/create, just saw some consistency moving toward enabled and it made sense to me. Feel free to change it.

jsulinski commented:
@icereval Do you want to add watchNamespace too? I haven't had time to test this yet but it would be good to add. Thanks!

mgoodness commented:
I can take care of watchNamespace, too, since it's RBAC-related. Will ping you all when I think it's ready.

jsulinski committed:
Add RBAC support

mgoodness committed:
Fix README

mgoodness committed:
Restrict namespace RBAC if scoped

mgoodness commented:
I think this will do it. RBAC for `--watch-namespace` is a bit [funky](https://github.com/kubernetes/charts/pull/1235/files#diff-a1944ae6cbf0a2ce4a54e7c531d44537R23). If `controller.scope.enabled=true` and `controller.scope.namespace` is set (meaning we're watching a namespace that *isn't* `.Release.Namespace`), we'll add `get namespaces` for that namespace to the ClusterRole. If `controller.scope.enabled=true` and `controller.scope.namespace` is *not* set, then we don't add `get namespaces` to the ClusterRole and simply rely on permission granted by the Role.

I've not actually tested this, but wanted to document it before I shut down for the night.

jsulinski commented:
Thanks a lot! I tested this and it works for my albeit limited use case. I have not tested namespaces, however. 

The design you described for clusterRole makes sense to me, but I would take that with a grain of salt. :)

mgoodness committed:
Tweak RBAC

mgoodness committed:
Bump chart versions

mgoodness commented:
/ok-to-test

mgoodness commented:
/lgtm

jsulinski commented:
Awesome. Thanks a lot for helping get this over the finish line! Really appreciate it.
